pkg/controllers/nodeclaim/underutilization/controller.go

```diff
@@ -0,0 +1,450 @@
+
+	if allExistingAreSpot &&
+		results.NewNodeClaims[0].Requirements.Get(v1beta1.CapacityTypeLabelKey).Has(v1beta1.CapacityTypeSpot) {
+		return c.validateSpotToSpotConsolidation(ctx, candidates, results, candidatePrice)
+	}
+
+	var incompatibleMinReqKey string
+	// filterByPriceWithMinValues returns the instanceTypes that are lower priced than the current candidate and the requirement for the NodeClaim that does not meet minValues.
+	// If we use this directly for spot-to-spot consolidation, we are bound to get repeated consolidations because the strategy that chooses to launch the spot instance from the list does
+	// it based on availability and price which could result in selection/launch of non-lowest priced instance in the list. So, we would keep repeating this loop till we get to lowest priced instance
+	// causing churns and landing onto lower available spot instance ultimately resulting in higher interruptions.
+	results.NewNodeClaims[0].NodeClaimTemplate.InstanceTypeOptions, incompatibleMinReqKey, _ =
+		disruptionutils.FilterByPriceWithMinValues(results.NewNodeClaims[0].InstanceTypeOptions, results.NewNodeClaims[0].Requirements, candidatePrice)
+
+	if len(results.NewNodeClaims[0].NodeClaimTemplate.InstanceTypeOptions) == 0 {
+		if len(candidates) == 1 {
+			if len(incompatibleMinReqKey) > 0 {
+				c.recorder.Publish(disruptionevents.Unconsolidatable(candidates[0].Node, candidates[0].NodeClaim, fmt.Sprintf("minValues requirement is not met for %s", incompatibleMinReqKey))...)
+			} else {
+				c.recorder.Publish(disruptionevents.Unconsolidatable(candidates[0].Node, candidates[0].NodeClaim, "Can't replace with a cheaper node")...)
+			}
+		}
+		return false, nil
+	}
+	return true, nil
+}
+
+// propagates the consolidation decision to cluster state and also marks all
+// associated nodeclaims as underutilized
+func (c *Controller) setNodeClaimsUnderutilized(ctx context.Context, setTrue bool, commandID types.UID, candidates ...*state.StateNode) error {
+	for i := range candidates {
+		nc := candidates[i].NodeClaim
+		cond := nc.StatusConditions().GetCondition(v1beta1.Underutilized)
+		// If we want to set the condition and it's already true
+		// or if we don't want to set the condition and it isn't true
+		// e.g. (T && T) || (F && F), then do nothing
+		if (setTrue && cond.IsTrue()) || (!setTrue && !cond.IsTrue()) {
+			continue
+		}
+		deepCopy := nc.DeepCopy()
+		if setTrue {
+			nc.StatusConditions().SetCondition(apis.Condition{
+				Type:    v1beta1.Underutilized,
+				Status:  v1.ConditionTrue,
+				Message: (string)(commandID),
+			})
+		} else {
+			err := nc.StatusConditions().ClearCondition(v1beta1.Underutilized)
+			if err != nil {
+				return fmt.Errorf("clearing underutilized condition from nodeclaim %q", nc.Name)
+			}
+		}
+		if err := c.kubeClient.Status().Patch(ctx, nc, client.MergeFrom(deepCopy)); err != nil {
```

Comment:
I've run into some nasty bugs patching status conditions from multiple concurrent controllers, and blowing away the data. CRDs don't support strategic merge. You may want to GET+Patch

Reply:
Depending on the changes to status we want to make, I wonder if we could use `application/json-patch+json`.

See https://kubernetes.io/docs/reference/using-api/api-concepts/#patch-and-apply for the available choices.